sql/expression/arithmetic.go

```diff
@@ -0,0 +1,1386 @@
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Int16:
+		l, r, err := toInt16(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Int32:
+		l, r, err := toInt32(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Int64:
+		l, r, err := toInt64(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Uint:
+		l, r, err := toUint(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Uint8:
+		l, r, err := toUint8(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Uint16:
+		l, r, err := toUint16(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Uint32:
+		l, r, err := toUint32(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Uint64:
+		l, r, err := toUint64(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Float32:
+		l, r, err := toFloat32(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+
+	case reflect.Float64:
+		l, r, err := toFloat64(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l / r, nil
+	}
+
+	return nil, ErrUnableToCast.New(lval, rval)
+}
+
+func bitAnd(lval, rval interface{}) (interface{}, error) {
+	kind, err := commonKind(lval, rval)
+	if err != nil {
+		return nil, err
+	}
+
+	switch kind {
+	case reflect.Int:
+		l, r, err := toInt(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Int8:
+		l, r, err := toInt8(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Int16:
+		l, r, err := toInt16(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Int32:
+		l, r, err := toInt32(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Int64:
+		l, r, err := toInt64(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Uint:
+		l, r, err := toUint(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Uint8:
+		l, r, err := toUint8(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Uint16:
+		l, r, err := toUint16(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Uint32:
+		l, r, err := toUint32(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+
+	case reflect.Uint64:
+		l, r, err := toUint64(lval, rval)
+		if err != nil {
+			return nil, err
+		}
+		return l & r, nil
+	}
+
+	return nil, ErrUnableToCast.New(lval, rval)
+}
+
+func bitOr(lval, rval interface{}) (interface{}, error) {
+	kind, err := commonKind(lval, rval)
+	if err != nil {
+		return nil, err
+	}
+
+	switch kind {
+	case reflect.Int:
```

Comment:
Yeah, this may be a little too much of an overkill for this. So is using reflection for this, as we already have the types of the operands.

I think is enough to use the following functions: https://github.com/src-d/go-mysql-server/blob/master/sql/type.go#L652-L670 to check which type it is and just treat the following four cases:
- int + int (independently of which size): treat as int64
- uint + uint (independently of which size): treat as uint64
- uint + int (independently of which size): treat as uint64
- anything else (independently of which size): float64

This would be simpler while still correct.

Reply:
I used `reflect`, because I needed all basic types. `sql/type.go` are tiny subset, moreover you have to rely on what `Type()` function returns for your left/right what's not always true, because you can pass any value, but your `Type` may return something different (e.g. because in Type you don't want to re-evaluate expressions).

Anyway if we all agree that 64-bit numbers as a result are good enough for all arithmetic operations (and does not matter what's the input's size) then it's fine. I'll refine PR.